wrappers/python/openmm/app/internal/xtc_utils/src/xtc.cpp

```diff
@@ -0,0 +1,326 @@
+        fprintf(stderr, "xtc_read(): could not get natoms\n");
+        return -1;
+    }
+    return natoms;
+}
+
+int xtc_nframes(char* filename) {
+    int natoms = 0;
+    int nframes = 0;
+    rvec* p = NULL;
+    XDRFILE* xd = NULL;
+    float time;
+    int step;
+    float prec;
+    matrix box;
+    if (exdrOK != read_xtc_natoms(filename, &natoms)) {
+        fprintf(stderr, "xtc_read(): could not get natoms\n");
+        return -1;
+    }
+    if (!natoms) {
+        fprintf(stderr, "xtc_read(): natoms is 0\n");
+        return -1;
+    }
+    xd = xdrfile_open(filename, "r");
+    if (NULL == xd) {
+        fprintf(stderr, "xtc_read(): could not open file\n");
+        return -1;
+    }
+    p = (rvec*)malloc(sizeof(rvec) * natoms);
+    int retval = 0;
+    uint64_t offset = 0;
+    while (exdrOK == (retval = read_xtc(xd, natoms, &step, &time, box, p, &prec))) {
+        nframes++;
+        offset = ftell(xd->fp);
+    }
+    condfree((void*)p);
+    p = NULL;
+    xdrfile_close(xd);
+    if (retval == exdr3DX) {
+        fprintf(stderr, "xtc_read(): XTC file is corrupt\n");
+        return -1;
+    }
+    return nframes;
+}
+
+void xtc_read_new(char* filename, float* coords_arr, float* box_arr, float* time_arr, int* step_arr, int natoms, int nframes) {
+    rvec* p = NULL;
+    XDRFILE* xd = NULL;
+    float time;
+    int step;
+    float prec;
+    matrix box;
+    int nf3 = nframes * 3; // Precalculate 3 * nframes for the coordinate lookup macro
+    if (!natoms) {
+        fprintf(stderr, "xtc_read(): natoms is 0\n");
+        return;
+    }
+    xd = xdrfile_open(filename, "r");
+    if (NULL == xd) {
+        fprintf(stderr, "xtc_read(): could not open file\n");
+        return;
+    }
+    p = (rvec*)malloc(sizeof(rvec) * natoms);
+    int retval = 0;
+    int fidx = 0;
+    int _natoms_garbage = 0;
+    int xidx, yidx, zidx, aidx;
+    while (exdrOK == (retval = read_xtc(xd, _natoms_garbage, &step, &time, box, p, &prec))) {
+        time_arr[fidx] = time;
+        step_arr[fidx] = step;
+        for (int i = 0; i < 3; i++) {
+            for (int j = 0; j < 3; j++) {
+                box_arr[fidx + (3 * i + j) * nframes] = box[i][j];
+            }
+        }
+        for (aidx = 0; aidx < natoms; aidx++) {
+            xidx = Xf(aidx, fidx, nframes);
+            yidx = Yf(xidx, nframes);
+            zidx = Zf(yidx, nframes);
+            coords_arr[xidx] = p[aidx][0];
+            coords_arr[yidx] = p[aidx][1];
+            coords_arr[zidx] = p[aidx][2];
+        }
+        fidx++;
+    }
+    condfree((void*)p);
+    p = NULL;
+    xdrfile_close(xd);
+    if (retval == exdr3DX) {
+        fprintf(stderr, "xtc_read(): XTC file is corrupt\n");
+    }
+}
+
+struct XTC_frame* xtc_read(char* filename, int* natoms, int* nframes, double* dt, int* dstep) {
+    struct XTC_frame* frames = NULL;
+    *natoms = 0;
+    *nframes = 0;
+    rvec* p = NULL;
+    XDRFILE* xd = NULL;
```

Comment:
This is another C-ism that can be handled better in C++.  You declare all your variables at the top of the function and initialize them to meaningless values.  Later on you assign the real values to them.  This creates a risk of bugs: there's a range of lines where the variable is declared and the compiler will let you use it, but any attempt to do so will probably create a bug.  It's better not to declare it until you're ready to assign the real value:

```c++
XDRFILE* xd = xdrfile_open(filename, "r");
```

That way, any attempt to reference `xd` before its value has been assigned will produce a compiler error.